File: brightspark.js
import fs from "fs";

// Custom Parse function
function parseCSV(content) {
  const lines = content.trim().split(/\r?\n/);
  const headers = lines.shift().split(",").map((h) => h.trim());

  // Process each line into an object
  return lines.map((line) => {
    const parts = [];
    let current = "";
    let in_quotes = false;

    // Handle quotes and commas inside quoted fields
    for (let i = 0; i < line.length; i++) {
      const char = line[i];
      if (char === '"' && line[i + 1] === '"') {
        current += '"';
        i++;
      } else if (char === '"') {
        in_quotes = !in_quotes;
      } else if (char === "," && !in_quotes) {
        parts.push(current);
        current = "";
      } else {
        current += char;
      }
    }

    parts.push(current);

    // Map the cell values to header names
    const record = {};
    headers.forEach((h, i) => {
      record[h] = parts[i]?.trim() ?? "";
    });

    return record;
  });
}

// Main 
function main() {
  // Exit if argument is missing
  if (process.argv.length !== 3) {
    console.error("Usage: node brightspark.js <input_file.csv>");
    process.exit(1);
  }

  // Input file
  const filePath = process.argv[2];

  // Try reading the file
  let data;
  try {
    data = fs.readFileSync(filePath, "utf8");
  } catch (err) {
    console.error(`Error: Cannot read file '${filePath}'.`);
    process.exit(1);
  }

  // Parse CSV
  const records = parseCSV(data);

  // If CSV file is empty
  if (records.length === 0) {
    console.log("No records found in file.");
    process.exit(0);
  }

  // Convert numeric fields
  for (const r of records) {
    r.division = parseInt(r.division);
    r.points = parseInt(r.points);

    if (isNaN(r.division)) {
      console.error("Error: divisions must be integers.");
      process.exit(1);
    }
    else if (isNaN(r.points)) {
      console.error("Error: points must be integers.");
      process.exit(1);
    }
  }

  // Sort records by ascending division, descending points
  const sorted = records.sort((a, b) => {
    if (a.division !== b.division) return a.division - b.division;
    return b.points - a.points;
  });

  // Top 3 records
  const topThree = sorted.slice(0, 3);

  // YAML Output
  console.log("records:");
  for (const r of topThree) {
    const name = `${r.firstname} ${r.lastname}`;
    const details = `In division ${r.division} from ${r.date} performing ${r.summary}`;
    console.log(`- name: ${name}`);
    console.log(`  details: ${details}`);
  }
}

// Driver
main();

File: brightspark.py
import yaml
import csv
import sys

def main():
    # Exit if argument is missing
    if len(sys.argv) != 2:
        print("Usage: python3 brightspark.py <input_file.csv>")
        sys.exit(1)

    # Get input file from command line argument
    file = sys.argv[1]

    # Try to open and parse CSV file
    try:
        with open(file, newline = '') as csvfile:
            # Read first row as header. Use column names as dict keys for each row
            reader = csv.DictReader(csvfile)
            records = list(reader)
    except FileNotFoundError:
        print(f"Error: File '{file}' not found.")
        sys.exit(1)

    # If CSV file is empty
    if not records:
        print("No records in file.")
        sys.exit(0)

    # Convert numeric fields to integers for sorting
    for idx, r in enumerate(records):
        try:
            r["division"] = int(r["division"])
            r["points"] = int(r["points"])
        except KeyError as error:
            # Missing expected column in the CSV header
            print(f"Error: CSV is missing column: {error}")
            sys.exit(1)
        except ValueError as error:
            # Non-integer value in numeric column
            print(f"Error: Invalid numeric value in row {idx + 1}: {error}")
            sys.exit(1)

    # Sort records by ascending division, descending points
    sorted_records = sorted(records, key = lambda x: (x["division"], -x["points"]))

    # Top 3 records
    top_three = sorted_records[:3]

    # YAML Output
    yaml_records = []
    for record in top_three:
        yaml_records.append({
            "name": f"{record['firstname']} {record['lastname']}",
            "details": f"In division {record['division']} from {record['date']} performing {record['summary']}"
        })

    output = {"records": yaml_records}

    print(yaml.dump(output, sort_keys = False).strip())

# Driver
if __name__ == "__main__":
    main()
